lib/ActivityStreams.js

```diff
@@ -0,0 +1,144 @@
+  this._setupListeners();
+  NewTabURL.override(this.pageURL);
+};
+
+ActivityStreams.prototype = {
+
+  pageMod: null,
+
+  /**
+   * Send a message to a worker
+   */
+  send(actionName, msg, worker) {
+    worker.port.emit("addon-to-content", {
+      type: actionName,
+      data: msg
+    });
+  },
+
+  /**
+   * Broadcast a message to all workers
+   */
+  broadcast(actionName, msg) {
+    for (let worker of this.workers) {
+      this.send(actionName, msg, worker);
+    }
+  },
+
+  /**
+   * Responds to places requests
+   */
+  _respondToPlacesRequests(msgName, params) {
+    let {msg, worker} = params;
+    PlacesProvider.links.getTopFrecentSites(msg.data).then(links => {
+      this.send("RECEIVE_TOP_FRECENT_SITES", links, worker);
+    });
+  },
+
+
+  /**
+   * Broadcast places changes to pages
+   */
+  _handlePlacesChanges(eventName, data) {
+    this.broadcast("RECEIVE_PLACES_CHANGES", {
+      "event": eventName,
+      "data": data
+    });
+  },
+
+  /**
+   * Sets up various listeners for the pages
+   */
+  _setupListeners() {
+    PlacesProvider.links.on("deleteURI", this._handlePlacesChanges.bind(this));
+    PlacesProvider.links.on("clearHistory", this._handlePlacesChanges.bind(this));
+    PlacesProvider.links.on("linkChanged", this._handlePlacesChanges.bind(this));
+    PlacesProvider.links.on("manyLinksChanged", this._handlePlacesChanges.bind(this));
+
+    this.on("REQUEST_TOP_FRECENT_SITES", this._respondToPlacesRequests.bind(this));
+  },
+
+  /**
+   * Turns off various listeners for the pages
+   */
+  _removeListeners() {
+    PlacesProvider.links.off("deleteURI", this._handlePlacesChanges);
+    PlacesProvider.links.off("clearHistory", this._handlePlacesChanges);
+    PlacesProvider.links.off("linkChanged", this._handlePlacesChanges);
+    PlacesProvider.links.off("manyLinksChanged", this._handlePlacesChanges);
+
+    this.off("REQUEST_TOP_FRECENT_SITES", this._respondToPlacesRequests);
+  },
+
+  /**
+   * Sets up communications with the pages
+   */
+  _setupPageMod() {
+    this.pagemod = new PageMod({
+      include: [this.pageURL],
+      contentScriptFile: data.url("content-bridge.js"),
+      contentScriptWhen: "start",
+      attachTo: "top",
+      onAttach: worker => {
+        if (!this.workers.has(worker)) {
+          this.workers.add(worker);
+        }
+        worker.port.on("content-to-addon", msg => {
+          if (!msg.type) {
+            console.warn(`ActivityStreams.dispatch error: unknown message type`);
+            return;
+          }
+          if (msg.type === "pagehide") {
+            console.debug(`ActivityStreams.pagemod unloading worker`);
+            this.workers.delete(worker);
+          }
+          this.emit(msg.type, {msg:msg, worker:worker});
+        });
+      }
+    });
+  },
+
+  /**
+   * Unload the application
+   */
+  unload(reason) { // eslint-disable-line no-unused-vars
+    console.debug(`ActivityStreams.unload on ${reason}`);
+
+    if(reason == "shutdown") {
```

Comment:
Q: if there might be other reason in the future, this might be better as switch - then you can handle all cases explicitly?
```JS
switch (reason){
case "shutdown":
  console.warn("Not implemented");
  break;
case "uninstall":
case "upgrade":
case "downgrade":
 this.workers.clear();
}
```

Or just:
```JS
switch (reason){
case "shutdown":
  console.warn("Not implemented");
  break;
default:
 this.workers.clear();
}
```